### DISCUSSION TITLE
Logic behind op.#Read? (does vela read the value on every Reconciliation cycle or only once)

### DISCUSSION THREAD

**Discussion by krikeri** [kubevela/kubevela#4681] [Q&A] [answered]
I am currently using op.#Read, the built-in workflow operator to read a config map from the cluster and use the value from that config map to render some other elements with Vela. 

What I noticed is that after the first time the workflow has finished and the value of the config map is changed, Vela is not reading the change anymore from the config map, and still provides the first value that was found. 

Does anyone know if that is by default or is something not working correctly in this case?

**Comment by wonderflow** [comment, marked as answer]
Yes. This is by default. After finished workflow won't be effected unless Application CR changes. We're launching a kube trigger project for this case, to watch outside resource and trigger event.  https://github.com/kubevela/kube-trigger

**Reply by krikeri** [reply]
@wonderflow thank you very much for the clarification and the tool suggestion. 
If I understood correctly the kube-trigger is changing the value of the "app.oam.dev/publishVersion" annotation by increasing it every time the value of the config map or another resource that is being monitored changes. 

Is the change on the annotation enough or does kube-trigger tool also modify something else in the background?

I am currently using something similar to the tool you created which is argo-events.
If the "app.oam.dev/publishVersion" annotation is changed manually or by using the kube-trigger/argo-events/third-party-tool I expect the result to be the same. Would you agree with my statement?

**Reply by wonderflow** [reply]
Yes, just change the annotation  "app.oam.dev/publishVersion"  is enough, that is the only thing kube-trigger to do for update application.

**Reply by krikeri** [reply]
I tested this feature and it works as expected, when the publishVersion gets changed the workflow triggers again and gets the latest value from the ConfigMap. 
Unfortunately, what I noticed is that after setting this annotation any new application changes are not triggering anymore the regular workflow execution. 

Is there any way how to combine the two mechanics, having the workflow triggered by annotation and by regular application changes?

**Reply by wonderflow** [reply]
Oh, yes, marking the `app.oam.dev/publishVersion` will make the application fixed in the specified version even if the application spec changed, this can be helpful that the application changes will always under control.

There're two ways to trigger an application workflow to re-run:

1. Change the publishVersion annotation.
2. Change any of the spec while don't set the publishVersion annotations.

In your case, I think maybe you can have some spec changed everytime you want to trigger the workflow to re-run without setting the `app.oam.dev/publishVersion` annotation.

**Reply by krikeri** [reply]
Thank you for the clarification. Hopefully one last question regarding the topic. :)
Below you can find an example of how I am defining one of the applications.
 
```
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: blue-cluster-gm-app
  namespace: vela-apps
spec:
  components:
    - name: gm-event-gke-sa
      type: gke-sa
    - name: gm-gke-sa
      type: gke-sa
  workflow:
    steps:
      - name: deploy-group-management
        properties:
          clusterName: blue
        type: deploy-group-management@v9
```

I can't change the name and the type of the component for as long as the component definition did not change. Similarly for the workflow part. 
Are there fields / hidden fields that potentially could be changed and would still trigger the workflow execution?

**Reply by wonderflow** [reply]
you can define a parameter in your `gke-sa` componentdefinition while the parameter was not used anywhere. Just change that noop parameter in component properties.

**Reply by krikeri** [reply]
Thank you for the hint.
